crates/emulated/src/field_ops.rs

```diff
@@ -433,7 +435,11 @@ where
         if a.is_constant() && b.is_constant() {
             let a_int = BigInt::from(a);
             let b_int = BigInt::from(b);
-            let res_int = (a_int + b_int).rem(P::modulus());
+            let mut res_int = a_int - b_int;
+            while res_int < BigInt::zero() {
+                res_int += P::modulus();
+            }
+            res_int = res_int.rem(P::modulus());
```

Comment:
Instead of `while` loop, can `res_int` be calculated as follows. I think the loop may take a while (no pun intended) if `a_int` is much smaller than `b_int`. It's possible that `a_int` was derived from 4 limbs of 64 bits each and `b_int` was derived from 8 limbs of 64 bits each.

``` rust
let res_int  = if a_int > b_int {
    (a_int-b_int).rem(P::modulus())
} else {
    P::modulus() - (b_int-a_int).rem(P::modulus())
} 
```

Reply:
Apart from my above nit, the fix looks fine to me. Thanks for the detailed investigation and tests.

The missing `return Ok(())` in `assert_is_equal` is a bug in my code. So is the addition instead of subtraction in `sub_op`. Thanks for finding these. I should have added tests for the constant cases.

Reply:
The while loop taking a long time for differing limb counts is a good point, thanks for the catch!